File: ruby/154_find_minimum_in_rotated_sorted_array_two.rb
# @param {Integer[]} nums
# @return {Integer}


# 找到被旋转的数组中最小的，其实也就是找到数组的head, 和153相比，此时数组会出现重复元素
# 虽然有重复元素，但是使用之前的第一种方法，找到数组的首元素，也就是最小的元素，也就是可以的

# 同样的也可以使用 类似 153 的思路，只是此时需要过滤掉重复元素
def find_min(nums)
	return -1 if nums.empty?
	return nums[0] if nums.size == 1 || nums[0] < nums[nums.size-1]

	left, right = 0, nums.size-1
	while left <= right
		while left < right && nums[left+1] == nums[left]
			left += 1
		end
		return nums[left] if left == right
		return nums[left+1] if nums[left+1] < nums[left]

		if right > left && nums[right-1] == nums[right]
			right -= 1
		end
		return nums[right] if left == right
		return nums[right] if nums[right-1] > nums[right]

		left += 1
		right -= 1
	end

	-1
end


# 不过既然使用了二分，也是可以用mid的方法来写
def find_min_version2(nums)
	return -1 if nums.empty?
	return nums[0] if nums.size == 1 || nums[0] < nums[nums.size-1]

	res = nums[0]
	left, right = 0, nums.size-1
	while left < right-1
		mid = left + (right-left)/2
		if nums[left] < nums[mid]
			res = [res, nums[left]].min 
			left = mid+1
		elsif nums[left] > nums[mid]
			res = [res, nums[right]].min 
			right = mid
		else
			left += 1
		end
	end

	res = [res, nums[left]].min 
	res = [res, nums[right]].min
	res
end


nums = [1, 1]
puts find_min(nums)

File: cpp/154_find_minimum_in_rotated_sorted_array_two.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

// 给定rotated sorted array，数组中可能存在重复元素，要求寻找到最小值返回；
// 1 7 7 7 7
// 7 1 7 7 7
// 1> 从左右开始遍历，找到逆序的返回，不过首先需要过滤掉重复元素；
// 2> 二分，因为存在重复，需要进行递归；
class Solution {
public:
    int findMin(vector<int>& nums) {
		if (nums.empty()) { return -1; }
		// sorted array
		if (nums.size() == 1 || (nums[0] < nums[nums.size()-1])) { return nums[0]; }

		int left = 0, right = nums.size()-1;
		while (left <= right) {
			while (left < right && nums[left+1] == nums[left]) { ++left; }
			if (left == right) { return nums[left]; }
			if (nums[left+1] < nums[left]) { return nums[left+1]; }

			while (right > left && nums[right-1] == nums[right]) { --right; }
			if (left == right) { return nums[left]; }
			if (nums[right-1] > nums[right]) { return nums[right]; }

			++left; --right;
		}

		return -1;
    }

	int helper(vector<int>& nums, int left, int right) {
		if (left == right) { return nums[left]; }
		int mid = left + (right - left) / 2;
		int l = 0, r = 0;

		if (nums[left] < nums[mid]) {
			l = nums[left];
		} else {
			l = helper(nums, left, mid);
		}
		if (nums[mid+1] < nums[right]) {
			r = nums[mid+1];
		} else {
			r = helper(nums, mid+1, right);
		}

		return min(l, r);
	}

	int findMin_bs(vector<int>& nums) {
		if (nums.empty()) { return -1; }
		// sorted array
		if (nums.size() == 1 || (nums[0] < nums[nums.size()-1])) { return nums[0]; }

		int left = 0, right = nums.size()-1;
		return helper(nums, left, right);
	}
};
